src/http/params.cr

```diff
@@ -78,6 +79,172 @@ module HTTP
+    #
+    # ```
+    # params.fetch("email", "none@example.org")                # => "john@example.org"
+    # params.fetch("non_existent_param", "default value")      # => "default value"
+    # ```
+    def fetch(name, default)
+      raw_params.fetch(name, [default]).first
+    end
+
+    # Returns first value for specified param name. Fallbacks to return value
+    # of provided block.
+    #
+    # ```
+    # params.fetch("email") { raise InvalidUser("email is missing") }    # InvalidUser "email is missing"
+    # params.fetch("non_existent_param") { "default computed value" }    # => "default computed value"
+    # ```
+    def fetch(name, &block : -> String)
+      raw_params.fetch(name) { [block.call] }.first
+    end
+
+    # Appends new value for specified param name. Creates param when there was
+    # no such param.
+    #
+    # ```
+    # params.add("item", "keychain")
+    # params.fetch_all("item")        # => ["pencil", "book", "workbook", "keychain"]
+    # ```
+    def add(name, value)
+      raw_params[name] ||= [] of String
+      raw_params[name] = [] of String if raw_params[name] == [""]
+      raw_params[name] << value
+    end
+
+    # Sets all values for specified param name at once.
+    #
+    # ```
+    # params.set_add("item", ["keychain", "keynote"])
+    # params.fetch_all("item")        # => ["keychain", "keynote"]
+    # ```
+    def set_all(name, values)
+      raw_params[name] = values
+    end
+
+    # Allows to iterate over all name-value pairs.
+    #
+    # ```
+    # params.each do |name, value|
+    #   puts "#{name} => #{value}"
+    # end
+    #
+    # # Outputs:
+    # # email => john@example.org
+    # # item => keychain
+    # # item => keynote
+    # ```
+    def each
+      raw_params.each do |name, values|
+        values.each do |value|
+          yield(name, value)
+        end
+      end
+    end
+
+    # Deletes first value for provided param name. If there are no values left,
+    # deletes param itself. Returns deleted value.
+    #
+    # ```
+    # params.delete("item")     # => "keychain"
+    # params.fetch_all("item")  # => ["keynote"]
+    #
+    # params.delete("item")     # => "keynote"
+    # params["item"]            # KeyError
+    #
+    # params.delete("non_existent_param")  # KeyError
+    # ```
+    def delete(name)
+      value = raw_params[name].shift
+      raw_params.delete(name) if raw_params[name].size == 0
+      value
+    end
+
+    # Deletes all values for provided param name. Returns array of deleted
+    # values.
+    #
+    # ```
+    # params.delete_all("comments")
+    # params.has_key?("comments")     # => false
+    # ```
+    def delete_all(name)
+      raw_params.delete(name)
+    end
+
+    # Serializes to string representation as http url encoded form
+    #
+    # ```
+    # params.to_s     # => "item=keychain&item=keynote&email=john@example.org"
+    # ```
+    def to_s(io)
+      io << HTTP::Params.build do |builder|
```

Comment:
We could make builder accept an `IO` in the constructor (defaulting to StringIO.new) so we write directly to it without creating intermediate strings.